### PR TITLE
Add ruby-reek support

### PR DESCRIPTION
**Pull request by franciscoj**
Hi!

I've been trying to add support for reek to the ruby parsers https://github.com/troessner/reek

I managed to make it work by adding the linter to my emacs dotfiles. It shows the errors as expected. ~~However I cannot make it pass the test. This is the first time I write some elisp and I'm a bit lost. Would anybody mind giving some help with it? I would really appreciate it :)~~ The tests are passing on the CI, but I cannot manage to make `rake check:language[ruby]` work on my machine. Don't know if it can be something relevant or not.

The error I get is:

``` lisp
(ert-test-failed
  ((should
    (equal expected flycheck-current-errors))
  :form
  (equal
    ([cl-struct-flycheck-error #<killed buffer> ruby-reek "/my-flycheck-working-copy/test/resources/language/ruby/smell-warnings.rb" 1 nil "has no descriptive comment" warning "IrresponsibleModule"]
    [cl-struct-flycheck-error #<killed buffer> ruby-reek "/my-flycheck-working-copy/test/resources/language/ruby/smell-warnings.rb" 2 nil "doesn't depend on instance state (maybe move it to another class?)" warning "UtilityFunction"]
    [cl-struct-flycheck-error #<killed buffer> ruby-reek "/my-flycheck-working-copy/test/resources/language/ruby/smell-warnings.rb" 3 nil "calls object.save 2 times" warning "DuplicateMethodCall"]
    [cl-struct-flycheck-error #<killed buffer> ruby-reek "/my-flycheck-working-copy/test/resources/language/ruby/smell-warnings.rb" 4 nil "calls object.save 2 times" warning "DuplicateMethodCall"])
    nil)
  :value nil :explanation
  (different-types
    ([cl-struct-flycheck-error #<killed buffer> ruby-reek "/my-flycheck-working-copy/test/resources/language/ruby/smell-warnings.rb" 1 nil "has no descriptive comment" warning "IrresponsibleModule"]
    [cl-struct-flycheck-error #<killed buffer> ruby-reek "/my-flycheck-working-copy/test/resources/language/ruby/smell-warnings.rb" 2 nil "doesn't depend on instance state (maybe move it to another class?)" warning "UtilityFunction"]
    [cl-struct-flycheck-error #<killed buffer> ruby-reek "/my-flycheck-working-copy/test/resources/language/ruby/smell-warnings.rb" 3 nil "calls object.save 2 times" warning "DuplicateMethodCall"]
    [cl-struct-flycheck-error #<killed buffer> ruby-reek "/my-flycheck-working-copy/test/resources/language/ruby/smell-warnings.rb" 4 nil "calls object.save 2 times" warning "DuplicateMethodCall"])
    nil)))
```

Would there be something else needed for this? Any documentation or other tests which would help?

### CODE REVIEW

File: flycheck.el [hunk 1]
```diff
@@ -7884,6 +7885,21 @@ See URL `http://jruby.org/'."
...
+(flycheck-def-config-file-var flycheck-reekrc ruby-reek "config.reek"
+  :safe #'stringp)
+
+(flycheck-define-checker ruby-reek
+  "A Ruby smeel checker using reek
+
+See URL `https://github.com/troessner/reek'."
+  :command ("reek" "--format=xml"
+            (config-file "--config" flycheck-reekrc)
+            source-original)
```
Comment (lunaryorn) [inline flycheck.el]: Why do we need `source-original` here?

**Comment by lunaryorn** [comment]
Thank you for this contribution.  I do not know why the tests failed, but the error indicates that no errors were parsed.  Note that we do not run syntax checker tests on Travis CI.

I can try myself, but not before next week.  Please be patient.

File: flycheck.el (hunk 1 above)
Comment (franciscoj) [inline flycheck.el]: Because I didn't read all the documentation to know how to do a checker. Sorry. I simplified the definition I hope this is better.

**Comment by franciscoj** [comment]
> Thank you for this contribution. I do not know why the tests failed, but the error indicates that no errors were parsed. Note that we do not run syntax checker tests on Travis CI.
> 
> I can try myself, but not before next week. Please be patient.

Thanks a lot for your reply. I'll wait, no problem :) I'll try to make the test pass on my own in the meantime. If you need anything regarding this, don't hesitate to contact.

**Comment by csaura** [comment]
Looks awesome! Thanks @franciscoj

File: flycheck.el [hunk 2]
```diff
@@ -232,6 +232,7 @@ attention to case differences."
     ruby-rubylint
     ruby
     ruby-jruby
+    ruby-reek
```
Comment (lunaryorn) [inline flycheck.el]: Why do you register this syntax checker after all other Ruby checkers?  That's also the reason why the test fails: it runs with the plain `ruby` checker which finds no errors of course.

How did you image this checker to be used in relation to the other ruby checkers?

File: flycheck.el [hunk 3]
```diff
@@ -7884,6 +7885,20 @@ See URL `http://jruby.org/'."
   :modes (enh-ruby-mode ruby-mode)
   :next-checkers ((warning . ruby-rubylint)))
 
+(flycheck-def-config-file-var flycheck-reekrc ruby-reek "config.reek"
+  :safe #'stringp)
+
+(flycheck-define-checker ruby-reek
+  "A Ruby smell checker using reek
```
Comment (lunaryorn) [inline flycheck.el]: Please add missing punctuation.

**Comment by lunaryorn** [comment]
@franciscoj I've added a couple of remarks to the diff.  Most notably the relation between this checker and the other existing Ruby checkers is still very unclear to me, and also the reason why the tests don't work.

Please take a look.

**Comment by franciscoj** [comment]
Hi @lunaryorn,

> How did you image this checker to be used in relation to the other ruby checkers?
> Add a line note

I imagined it being used in conjunction with rubocop (or ruby-lint) since one provides stylish/linting checking, while the other (reek) provides smell checking. At least that's the way I use it.

> Why do you register this syntax checker after all other Ruby checkers?

I just wanted to add it as the least priority one.

Now that I give a second thought on this, I think I made it wrong. Maybe instead of having rubocop in `:next-checkers` for reek, it should be reek who is there for rubocop. Would that make sense?

Thanks **a lot** for your review.

**Comment by lunaryorn** [comment]
@franciscoj Probably, I'm really not sure.  Theoretically it'd make sense for reek to come first (as lints are more important than style issues), but essentially it comes down to the question whether reek handles syntax errors gracefully.

We currently run rubocop before ruby-lint because by the time we added both rubocop reported syntax errors as normal errors whereas ruby-lint just propagated parser exceptions which are impossible to parse reasonably.

**Comment by franciscoj** [comment]
@lunaryorn Any specific tests I can add to check the way reek checks syntax so that we know what to do?

**Comment by lunaryorn** [comment]
Open `test/resources/languages/ruby/syntax-error.rb` and run `M-x flycheck-compile RET ruby-reek`.  That'll show you what reek reports for syntax errors.

**Comment by franciscoj** [comment]
This is the result I get:

```
-*- mode: compilation; default-directory: "~/Documents/src/flycheck/test/resources/language/ruby/" -*-
Compilation started at Fri Mar  4 09:08:45

reek --format\=xml /path/to/src/flycheck/test/resources/language/ruby/syntax-error.rb
/path/to/src/flycheck/test/resources/language/ruby/syntax-error.rb:5:7: error: unexpected token tCONSTANT
/path/to/src/flycheck/test/resources/language/ruby/syntax-error.rb:5: puts "Here are the days"
/path/to/src/flycheck/test/resources/language/ruby/syntax-error.rb:5:       ^~~~              
/path/to/src/flycheck/test/resources/language/ruby/syntax-error.rb: Parser::SyntaxError: unexpected token tCONSTANT
<?xml version='1.0'?>
<checkstyle/>

Compilation finished at Fri Mar  4 09:08:45
```

**Comment by lunaryorn** [comment]
@franciscoj That's unfortunate; as you can see the syntax error is reported separately, and not in the XML output :disappointed: 

Would you mind to report this to `reek`?  Maybe they can fix that, and we could put `reek` first where it belongs?

Or maybe we can restructure the entire ruby chain, and put plain `ruby` first for syntax checking, then `reek` or `ruby-lint`, and finally `rubocop`.  What do you think?  I understand that Ruby's syntax error messages have quite improved in recent Ruby versions (quite unlike back then when we added `ruby` in version 1.9.x or so)…

**Comment by lunaryorn** [comment]
I'm inviting @bbatsov and @purcell, my Ruby experts of choice, for their feedback, as they contributed much to Flycheck's Ruby support :relaxed:

**Comment by bbatsov** [comment]
> I understand that Ruby's syntax error messages have quite improved in recent Ruby versions (quite unlike back then when we added ruby in version 1.9.x or so)…

Did they? :-) 

Btw, RuboCop will report exactly the same syntax errors as Ruby would (but the messages would definitely be nicer). As for the execution order - I can certainly imagine people using `reek` and `rubocop` together (although there will be significant overlap in the problems being reported). RuboCop checks way more things than most people are aware of. E.g. recent versions added performance checks as well.

**Comment by lunaryorn** [comment]
@bbatsov Oh, apparently not then :wink: I'm not at all folllowing Ruby's development :blush:

**Comment by franciscoj** [comment]
@lunaryorn would it be an option to have rubocop to run and then reek?

If not, I can totally take this to reek and try to get the issue fixed there and then try again.

@bbatsov @lunaryorn thanks a lot both of you for your help!

**Comment by franciscoj** [comment]
@lunaryorn I've been having another look into this. Would it be an option to redirect STDERR to `/dev/null` to ignore those syntax errors in reek? In that case 

By doing something like this `reek test/resources/language/ruby/syntax-error.rb  --format=xml 2> /dev/null` I get no syntax errors but I get the proper (empty) checkstyle xml output.

```xml
<?xml version='1.0'?>
<checkstyle/>
```

I'm also trying to have a look into reek to see if there's any possibility to redirect that inside the tool instead of doing it on flycheck and, in case it's  not possible add a flag for it.

**Comment by lunaryorn** [comment]
We do not support redirections in Flycheck, nor would we want to use them for this purpose, even if we did.

I would rather prefer if reek properly included syntax errors in its checkstyle output.  Anything else is really not an option for us.

**Comment by franciscoj** [comment]
So, to sum up. The way to go would be to have both the smell warnings and the syntax errors inside reek in checkstyle? or would it make sense to suppress the syntax errors in reek?

**Comment by lunaryorn** [comment]
Either way is ok as far as Flycheck is concerned, though the former would be preferable.    But printing exception messages in plain text to standard error even if the output format is explicitly chosen as XML is not particularly great.

We can cope with this in Flycheck by running rubocop first, which should catch most errors, but it's by far not ideal.  If we could get Reek fixed before merging this PR that'd be much better.

**Comment by franciscoj** [comment]
great, I'll see what I can do. Thanks a lot for your help!

**Comment by lunaryorn** [comment]
@franciscoj Did you manage to fix reek to properly report syntax errors?

**Comment by franciscoj** [comment]
@lunaryorn not yet, sorry :( I was wondering if maybe it is better to close this PR and try to reopen it once ti manage to do it. Seems fine?

**Comment by lunaryorn** [comment]
@franciscoj However you want.  If you feel that you don't have time for it, or don't want to do it anymore, you can close it, but if you still intend to work on this, no matter how long it might take, you may keep it open so that we know that you're still on it.

As far as we are concerned it's totally fine to keep pull requests open for a long time  :relaxed:

**Comment by franciscoj** [comment]
> As far as we are concerned it's totally fine to keep pull requests open for a long time

Perfect!! I'll keep working on it, I just didn't have time in the last weeks. Thanks for your help!

**Comment by franciscoj** [comment]
I'm finally not being able to do this, sorry :( should I close the PR? Or leave it open just in case somebody wants to continue with it?

**Comment by Simplify** [comment]
@franciscoj Leave it open for now.

**Comment by franciscoj** [comment]
@Simplify ok!

**Comment by Simplify** [comment]
New PR: https://github.com/flycheck/flycheck/pull/1244

**Pull request closed without merging**